src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +367,324 @@ public class SqlMutationEngine : IMutationEngine
+                            }
+
+                            transactionScope.Complete();
+                        }
+                    }
+
+                    // All the exceptions that can be thrown by .Complete() and .Dispose() methods of transactionScope
+                    // derive from TransactionException. Hence, TransactionException acts as a catch-all.
+                    // When an exception related to Transactions is encountered, the mutation is deemed unsuccessful and
+                    // a DataApiBuilderException is thrown
+                    catch (TransactionException)
                     {
-                        isUpdateResultSet = Convert.ToBoolean(isUpdateResultSetValue);
+                        throw _dabExceptionWithTransactionErrorMessage;
                     }
 
+                    Dictionary<string, object?> resultRow = upsertOperationResultSetRow.Columns;
+
                     // For MsSql, MySql, if it's not the first result, the upsert resulted in an INSERT operation.
-                    // Even if its first result, postgresql may still be an insert op here, if so, return CreatedResult
-                    if (!isUpdateResultSet ||
-                        (sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL &&
-                        PostgresQueryBuilder.IsInsert(resultRow)))
+                    // With postgresql, even if it is the first result, the upsert could have resulted in an INSERT. So, that condition is evaluated.
+                    if (sqlMetadataProvider.GetDatabaseType() is DatabaseType.PostgreSQL)
                     {
-                        string primaryKeyRoute = ConstructPrimaryKeyRoute(context, resultRow, sqlMetadataProvider);
-                        // location will be updated in rest controller where httpcontext is available
-                        return new CreatedResult(location: primaryKeyRoute, OkMutationResponse(resultRow).Value);
+                        isUpdateResultSet = !PostgresQueryBuilder.IsInsert(resultRow);
                     }
 
-                    // Valid REST updates return OkObjectResult
-                    return OkMutationResponse(resultRow);
-                }
-            }
-            else
-            {
-                DbResultSetRow? mutationResultRow = null;
+                    // When read permissions is configured without database policy, a subsequent select query will not be executed.
+                    // However, the read action could have include and exclude fields configured. To honor that configuration setup,
+                    // any additional fields that are present in the response are removed.
+                    if (isReadPermissionConfiguredForRole && !isDatabasePolicyDefinedForReadAction)
+                    {
+                        IEnumerable<string> allowedExposedColumns = _authorizationResolver.GetAllowedExposedColumns(context.EntityName, roleName, EntityActionOperation.Read);
```

Comment:
In `OperationMetadata`, the `AllowedExposedColumns` field is already a HashSet. So, the actual type of the `allowedExposedColumns` would be a HashSet even though the type is declared as `IEnumerable<string>`.  The type of `allowedExposedColumns` is declared as `IEnumerable<string>` because the method signature of `GetAllowedExposedColumns` specifies the return type as  `IEnumerable<string>`

Reply:
Thanks for clarifying. That's good to know. Also I had missed this in docs which explain (HashSet impl:
> If the type of source implements [ICollection<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1?view=net-7.0), the Contains method in that implementation is invoked to obtain the result. Otherwise, this method determines whether source contains the specified element.